### PR TITLE
Do Not Merge: Adding TTL to the workflow Execution

### PR DESCRIPTION
**Pull request by agautam478**
**What changed?**
The Workflow Executions will have a time to live attached to them when they get created. This means that after the execution is over and it's retention period is over it will be removed from the Execution table.

TTL value = workflow start-to-close timeout + domain retention (or, we could just pick 30 days here, there's no reason to get the exact retention, just an upper bound is fine) + 15 days of leeway

**Why?**
These changes have been made as part of clean up process so that the partition size doesn't become bloated after the executions have served their purpose.

**How did you test it?**
Testing locally and added a test case.

**Potential risks**
In a worse case scenario, executions that are not supposed to be removed can be deleted permanently.

**Release notes**
NA

**Documentation Changes**
NA

### CODE REVIEW

File: service/history/execution/context.go [hunk 1]
```diff
@@ -45,6 +45,7 @@ import (
 
 const (
 	defaultRemoteCallTimeout = 30 * time.Second
+	buffertimeforTTL         = 15
```
Comment (davidporter-id-au) [inline service/history/execution/context.go:48]: Would you mind changing this to be a value which indicates the value it's on, I assume it's days? 

`ttlBufferDays` might be one option?

File: common/persistence/dataManagerInterfaces.go [hunk 2]
```diff
@@ -872,6 +872,8 @@ type (
 		NewWorkflowSnapshot WorkflowSnapshot
 
 		DomainName string
+
+		TTL int64
```
Comment (Shaddoll) [inline common/persistence/dataManagerInterfaces.go:876]: nit: we need a better name for this. `TTLInSeconds` is a bit better

File: common/dynamicconfig/constants.go [hunk 3]
```diff
@@ -3871,6 +3879,11 @@ var BoolKeys = map[BoolKey]DynamicBool{
 		Description:  "Enable shardId metrics in persistence client",
 		DefaultValue: true,
 	},
+	EnableTTL: DynamicBool{
```
Comment (Shaddoll) [inline common/dynamicconfig/constants.go:3882]: We need a better name. e.g. EnableExecutionTTL indicating that it's for execution table

File: common/dynamicconfig/constants.go [hunk 4]
```diff
@@ -3871,6 +3879,11 @@ var BoolKeys = map[BoolKey]DynamicBool{
...
+		KeyName:      "enableTTL",
```
Comment (Shaddoll) [inline common/dynamicconfig/constants.go:3883]: the key name should follow the existing naming pattern.
We can name it as `system.enableExecutionTTL`

File: common/dynamicconfig/constants.go [hunk 5]
```diff
@@ -3871,6 +3879,11 @@ var BoolKeys = map[BoolKey]DynamicBool{
...
+		Description:  "EnableTTL is which domains are allowed to have workflow executions with a TTL",
+		DefaultValue: true,
```
Comment (Shaddoll) [inline common/dynamicconfig/constants.go:3885]: DefaultValue should be false

**Comment by agautam478** [comment]
Closing the PR as the CreateWorkflow flow is not the correct place to update the TTL. It keeps on become zero as the workflow data gets updated.
New PR: https://github.com/uber/cadence/pull/5243

**Pull request closed without merging**